WAVE: RF64, extended MIME type support, and bug fixes

david-russo opened the pull request:
- Added support for parsing and validating RF64 files
- Made WAVE parser more resilient to unexpected chunk data
- Improved reporting of WAVE codecs in WAVEFORMATEXTENSIBLE files
- Avoids reporting file format and MIME type until signatures have been verified
- Now reports extended MIME type information, e.g. `audio/vnd.wave; codec=1`, as per RFC 2361
- Subformat GUID's are now reported in their standard format, e.g. `00000001-0000-0010-8000-00AA00389B71`, instead of as an array of byte values
- Added check to verify the existence of Data chunks
- Added check to verify that Format chunks appear before Data chunks
- Expanded WAVE example corpora to cover more formats and errors
- Improved truncation detection and reporting
- Fixed erroneous reporting of Cue Point values
- Renamed "Cue" report property to "CuePoints"

Closes #184 and fixes #278.

david-russo committed:
WAVE: Add basic RF64 support and require Java 8

The WAVE module can now recognize, read, and validate RF64 files up to
9.22 exabytes in size, support chunks with extended chunk sizes, and
files with large sample counts.

david-russo committed:
WAVE: Improve WAVEFORMATEXTENSIBLE support

- The module can now recognize the subformat GUIDs for PCM and other
  WAVEFORMAT codecs, allowing them to be identified and interpreted
  correctly.
- Subformat GUIDs are now shown in their standard form instead of as
  arrays of bytes.
- Profile verification for BWF files has been simplified and some
  code comments clarified.

david-russo committed:
WAVE: Improve format identification

Setting the format and MIME type information has been moved to after
the format signature has been verified to prevent the module from
characterizing all files as WAVE files. The MIME type has also been
extended to include WAVE codec ID's per RFC 2361.

david-russo committed:
WAVE: Ensure whole chunks are read or passed over

Fixes an issue where unexpected null data at the end of a chunk could
leave the parser at an invalid position to read any subsequent chunks.

tledoux commented:
Hi, is the requirement of Java 8 mandatory ?
Jhove is a fundation piece of code for preservation ; so asking for Java 8 forces everything on top of it to be in Java 8. 

Maybe a requirement for java 7 would be enough. In any case, a quick survey is needed before going to java 8...

david-russo commented:
The requirement comes from the use of Java 8's new unsigned-operation methods, such as `Long.compareUnsigned()`. It could be removed if we re-implemented those functions locally, though I'd still encourage a move to Java 8: support from Oracle for 7 [ended in 2015](https://java.com/en/download/faq/java_7.xml), and is [ending in June this year](https://access.redhat.com/articles/1299013) for the OpenJDK. Java 8 has had almost 4 years worth of maintenance releases at this point, so I'd think it should be fairly well supported by now, but I certainly see the prudence in a survey given the infrastructure JHOVE might be embedded in, as you mentioned.

david-russo committed:
WAVE: Improve message accuracy and code style

david-russo committed:
JHOVE: Improve truncation detection when skipping bytes

We now verify the amount of bytes skipped using InputStream's
`available` method to see how many bytes it's possible to skip
beforehand. To support these changes, the `available` method has
been implemented for JHOVE's InputStream classes where required,
namely RAFInputStream.

david-russo committed:
WAVE: Expand validation tests

- Expanded WAVE example corpora to cover more tests
- New check to verify existence of Data chunks
- New check to verify Format chunks appear before Data chunks
- Improved truncation detection and reporting

david-russo committed:
WAVE: Fix Cue Point reporting

ChunkStart values were being reported as BlockStart values,
BlockStart values were being reported as SampleOffset values,
and SampleOffset values were missing from the report entirely.
All values are now reported correctly.

david-russo committed:
Revert WAVE module requirement for Java 8

The required Java 8 method `Long.compareUnsigned` has been copied into
the WAVE module where it was needed, with a caveat that it should be
replaced once Java 8 support has been approved.

david-russo commented:
The requirement for Java 8 has now been removed, reverting compatibility to Java 6.

tledoux commented:
Thanks. Very much appreciated

carlwilson committed:
Merge branch 'integration' into rf64-support

carlwilson commented on jhove-core/src/main/java/edu/harvard/hul/ois/jhove/RAFInputStream.java:
```diff
@@ -19,106 +21,117 @@
     }
 
-    private void init ()
+    private void init()
```
Super pernickety and I have no good answer but the constructor above is corrected with the open bracket alongside the declaration (correct IMO) but for init just the spacing is corrected. Sorry take that back, mis-read the above and spent too much time recently working through JHOVE's inconsistent spacing / formatting issues.

carlwilson commented on jhove-core/src/main/java/edu/harvard/hul/ois/jhove/handler/XmlHandler.java:
```diff
@@ -4355,21 +4355,21 @@ private void writeAESTimeRange (String baseIndent,
-                    ("tcf:frames", Integer.toString (duration.getFrames ())));
+			       Long.toString (duration.getSeconds ())));
+	    _writer.println (margn2 + element ("tcf:frames",
+			       Long.toString (duration.getFrames ())));
```
Mixed spacing by the looks

david-russo replied:
Oh yeah, the *Handler files are a mess with white-space inconsistencies, even with tabs and spaces on the same line at times, or switching every handful of lines. I originally attempted to fix it, but there was just too much.

The half tabs and half spaces is actually consistent with the code's surrounding lines, which you can see by slowly highlighting line 4370 above.

carlwilson commented on jhove-core/src/main/java/edu/harvard/hul/ois/jhove/handler/TextHandler.java:
```diff
@@ -1073,19 +1073,23 @@ private void writeAESTimeRange (String baseIndent,
-	_writer.println (margn3 + "NumberOfSamples: " + 
-			 Integer.toString (start.getSamples ()));
+        _writer.println (margn3 + "NumberOfSamples: " +
+                Long.toString (start.getSamples ()));
```
Mixed spacing on this and the next four changes from Integer to Long.

david-russo replied:
Yeah, I probably should have matched the spacing with the line directly above instead of the line beneath, but this is one of those areas that is already in a sea of inconsistencies.

carlwilson commented on jhove-core/src/main/java/edu/harvard/hul/ois/jhove/handler/XmlHandler.java:
```diff
@@ -4378,7 +4378,7 @@ private void writeAESTimeRange (String baseIndent,
 						    sampleAttrs));
 	    _writer.println (margn3 + element ("tcf:numberOfSamples",
-			       Integer.toString (duration.getSamples ()) ));
+			       Long.toString (duration.getSamples ()) ));
```
Mixed spacing.

david-russo replied:
A pre-existing issue.

carlwilson commented on jhove-modules/src/main/java/edu/harvard/hul/ois/jhove/module/wave/FormatChunk.java:
```diff
@@ -108,113 +135,157 @@ public boolean readChunk(RepInfo info) throws IOException {
+                mapLoc[0] = "LEFT";
+                mapLoc[1] = "RIGHT";
+                break;
+
```
Mixed spacing.

david-russo replied:
Looks the same to me... Should we not indent switch cases?

carlwilson commented on jhove-modules/src/main/java/edu/harvard/hul/ois/jhove/module/wave/FormatChunk.java:
```diff
@@ -108,113 +135,157 @@ public boolean readChunk(RepInfo info) throws IOException {
-            }
+                for (int i = 0; i < numChannels; i++) {
+                    mapLoc[i] = "UNKNOWN";
+                }
```
Mixed spacing.

carlwilson reviewed (changes requested):
We should chat briefly about JHOVE formatting.

carlwilson committed:
Merge branch 'integration' into rf64-support

carlwilson committed:
Merge branch 'integration' into rf64-support
